Add guide: Observe your API traffic with ngrok and Datadog (#1351)
ngrok/ngrok-docs: docs/guides/api-gateway/monitor-ngrok.mdx

- Click the **Replay** button and notice that the request is replayed from the ngrok host.

  ![Replay request](img/monitor-ngrok/replay.png)

- You can also replay a request with changes to alter any of the headers or POST data.

  ![Replay with changes](img/monitor-ngrok/replayChanges2.png)

Replaying requests is useful for debugging.
For example, you could find the request that caused an error in your API, deploy a fix for the API, and replay the request to confirm you've fixed the issue.

### Observe and debug Traffic Policy rules with request replay

You can also replay requests to test new [Traffic Policy](https://ngrok.com/docs/traffic-policy) rules.
ngrok applies the Traffic Policy rules in effect at the current time of replay, which may be different from the rules that were in effect when the original request was received.

This replay function lets you test how new Traffic Policy rules behave against real-world traffic from the past, offering one more way to test or debug them.

Since you came here from a previous API gateway guide, you've already seen how to set up the [`rate-limit` action](/docs/traffic-policy/actions/rate-limit).

> Now you can not only observe what happens to rate-limited requests, but use Traffic Inspector to dial in your threshold.

Reviewer: I suspect this phrasing "now you can not only" might be difficult for non-native English speakers to parse.

```suggestion
Now you can observe what happens to rate-limited requests and also use Traffic Inspector to dial in your threshold.
```

- **Step 3:** Set the time preference to **Past 1 Hour**.
- **Step 4:** Name the widget `Errors in the last hour`.

  ![Datadog edit widget](img/monitor-ngrok/datadogWidgetDetails.png)

Your new widget will be available in the dashboard, allowing your support staff to see instantly whether any errors have occurred in your API.

![Datadog custom dashboard](img/monitor-ngrok/datadogCustomDashboard.png)

Since the sample API never returns errors, an easy way to test the `Errors in the last hour` widget is to stop the sample API Docker container and then try to browse to the site on the public ngrok endpoint.

If you want to create widgets for other log information, you can see which fields are available by reading the JSON of any event you click on in the log inspector.

### Create a notification

To complete your monitoring system, you need to set up an alert pushed to your email or mobile app when an error occurs by adding a webhook integration to Datadog.
This allows your support team to receive error notifications instead of repeatedly checking the dashboard.

> Let's set up webhooks with [ntfy.sh](https://ntfy.sh), a free notification service, as a quick demonstration.

Reviewer: nitpick: prefer framing actions in terms of "you" rather than "we/us" (let's)

```suggestion
You can set up webhooks with the free notification service [ntfy.sh](https://ntfy.sh) as a quick demonstration.
```

- In the Datadog navigation panel, browse to the **Monitors > New monitor**. Choose **Logs**.

- In the monitor configuration, choose **Query** for your monitor scope, set the search query to `@http.status_code:500`, the time to the **last 1 hour**, and the notification message:

  ```
  @webhook-ntfy Your API has errors. Investigate at https://app.datadoghq.eu/dashboard.
  ```

  ![Datadog add webhook](img/monitor-ngrok/datadogConfigureMonitor.png)

  It's important that the `@webhook` name matches the webhook integration you created earlier.
  If the names don't match, the notification won't arrive.

- Browse to the [ngrok test channel on the ntfy website](https://ntfy.sh/ngrokTest).

- Click **Test notifications** at the bottom of the Datadog page and send an alert with **Run tests**.

- Notice that the alert appears on the ntfy page.
  You can install ntfy as a mobile app so that you are always aware of whether your API has errors.

  ![Datadog notify](img/monitor-ngrok/datadogNotify.png)

- Save the monitor.
  Now, after a delay and once you've refreshed the ngrok endpoint for your API, you will receive a notification if an error occurs.

> ## Next steps: What to monitor?

Reviewer: nitpick: feels strange having one section called "next steps" and another one called "what's next" - since the latter is fairly standard as a conclusion in the docs we could change this one

```suggestion
## What to monitor
```

Now that you've seen how to connect your API traffic to Datadog, explore logs, create a dashboard with custom widgets, and send alerts, let's discuss what exactly you should be monitoring.

There are three categories to consider: security, errors, and performance.

### Security

> You need to monitor the security of your infrastructure and any attempts to hack
> users' accounts.

Reviewer: ```suggestion
You need to monitor the security of your infrastructure and any attempts to hack users' accounts.
```

You should set up alerts for any changes to API keys, domains, and SSH certificates—anything that indicates attackers might gain access to your infrastructure.
Events like stopping and starting agents and tunnels are likely to happen frequently, so they don't need alerts (especially when running Kubernetes and multiple containers).
You still might want to add widgets to monitor these events on a dashboard, however, because surges in activity might indicate problems with your system
Additionally, you can add monitors that trigger alerts when activity passes a certain threshold.

In terms of user security, you need to monitor anything that indicates attackers might be accessing a user's account.
This includes multiple requests from IP addresses in different countries in a short time span (though this has become common with the increasing use of VPNs), a high rate of failed login attempts, or an anomalous pattern of request access.
To make a monitoring rule that recognizes anomalies, you first need to track the average number of requests the average user makes to various endpoints in your API.

### Errors

Errors are the simplest category to monitor.

> Generally, if your API returns status codes in the `500` range instead of the `200` range, you have some debugging and remedation to attend to.

Reviewer: ```suggestion
Generally, if your API returns status codes in the `500` range instead of the `200` range, you have some debugging and remediation to attend to.
```

Performance metrics need monitoring but don't need to trigger alerts unless your service experiences a sudden spike in usage or growth.
The most important things to monitor are uptime, latency (response time), and throughput.

Unlike with audit events, there is only one traffic event to monitor: [HTTP Request Complete](https://ngrok.com/docs/obs/events/reference/#http-request-complete).
This event, with multiple fields and the request and response bodies in the log, is enough to construct any metrics you need.

General statistics practices apply to metric monitoring.
Use medians instead of arithmetic averages, because extreme values disproportionately affect averages.

For latency, keep an eye on the median duration of a request and outliers—separate requests by endpoint to identify which endpoints take excessive time and could be improved.
For throughput, look at the median number of requests per minute.
For both metrics, also consider monitoring variance in addition to averages.
Even if an endpoint has a quick average response time, if the duration is highly variable, with many lower- and higher-than-average durations, your users will probably get annoyed.

## What's next?

> You've now not only set up ngrok as an API gateway with the agent or Kubernetes Operator, but also established fundamentals for observing it as a holistic system.

Reviewer: another opportunity to simplify the grammar

```suggestion
By the end of this guide, you've set up ngrok as an API gateway with the agent or Kubernetes Operator and established fundamentals for observing it as a holistic system.
```

through it with ngrok by viewing traffic reports, using error request replays,
and exporting logs and events to an external dashboard.

## What you'll need

- A [Datadog account](https://www.datadoghq.com/) and API key.
- A project with ngrok deployed as an API gateway. If you don't have one, you can use one of the following guides to get started:
  - [Get started with ngrok's API gateway](https://ngrok.com/docs/guides/api-gateway/get-started/)
  - [Deploy an API gateway with Kubernetes](https://ngrok.com/docs/guides/api-gateway/kubernetes/)
  - [Deploy a multicloud API gateway](https://ngrok.com/docs/guides/api-gateway/multicloud/)

## Options for monitoring your API gateway and services

You can monitor API services in two ways in ngrok: with the [Traffic
Inspector](/docs/obs/traffic-inspection/) and with [Events](/docs/obs/events).

<Tabs groupId="monitoring" queryString="type">
	<TabItem value="traffic-inspector" label="Traffic Inspector">
	
The Traffic Inspector is a filterable list of your API's requests and responses,
like structured logs in a human-readable format, available on the ngrok
dashboard. The Traffic Inspector is useful for viewing error details and for

> replaying requests to test new policies and bug fixes.

Author: Hey @sg-writer! Totally happy to do that, and I've added it to my to-do list, but it might take me a day or two to get back around to it.

> replaying requests to test new policies and bug fixes.

Reviewer: Thank you so much for doing this @joelhans !!
